docs/src/manual/exporting_to_jax.md

```diff
@@ -0,0 +1,144 @@
+# Exporting Lux Models to Jax (via EnzymeJAX & Reactant)
+
+!!! danger "Experimental"
+
+    This feature is experimental and is subject to change without notice. Additionally,
+    this feature currently requires some manual setup for interacting with Jax, which we are
+    working on improving.
+
+In this manual, we will go over how to export Lux models to StableHLO and use
+[EnzymeJAX](https://github.com/EnzymeAD/Enzyme-JAX) to run integrate Lux models with
+JAX. We assume that users are familiar with
+[Reactant compilation of Lux models](@ref reactant-compilation).
+
+```@example exporting_to_stablehlo
+using Lux, Reactant, Random
+
+const dev = reactant_device()
+```
+
+We simply define a Lux model and generate the stablehlo code using `Reactant.@code_hlo`.
+
+```@example exporting_to_stablehlo
+model = Chain(
+    Conv((5, 5), 1 => 6, relu),
+    MaxPool((2, 2)),
+    Conv((5, 5), 6 => 16, relu),
+    MaxPool((2, 2)),
+    FlattenLayer(3),
+    Chain(
+        Dense(256 => 128, relu),
+        Dense(128 => 84, relu),
+        Dense(84 => 10)
+    )
+)
+ps, st = Lux.setup(Random.default_rng(), model) |> dev;
+```
+
+Generate an example input.
+
+```@example exporting_to_stablehlo
+x = randn(Random.default_rng(), Float32, 28, 28, 1, 4) |> dev;
+```
+
+No instead of compiling the model, we will use `Reactant.@code_hlo` to generate the
+StableHLO code.
+
+```@example exporting_to_stablehlo
+hlo_code = @code_hlo model(x, ps, st)
+```
+
+Now we just save this into an `mlir` file. Remember that we only need to save the `@main`
+function and not the entire module.
```

Comment:
We should just export the entire module, in case some functions aren’t inlined

Reply:
I was getting 

```python
The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Cell In[4], line 15
     11 bias6_3 = jax.random.normal(jax.random.PRNGKey(0), (24, 24, 6, 4))
     13 # Run the exported Lux model
     14 print(
---> 15     jax.jit(run_lux_model)(
     16         x,
     17         weight1,
     18         bias1,
     19         weight3,
     20         bias3,
     21         weight6_1,
     22         bias6_1,
     23         weight6_2,
     24         bias6_2,
     25         weight6_3,
     26         bias6_3,
     27     )
     28 )

    [... skipping hidden 19 frame]

File ~/miniconda3/envs/enzymejax/lib/python3.11/site-packages/enzyme_ad/jax/primitives.py:809, in _enzyme_primal_lowering(ctx, source, fn, argv, out_shapes, lang, pipeline_options, *args_flat)
    806 for f in mod.regions[0].blocks[0]:
    807     fns.append(f.sym_name.value)
--> 809 name, nmod = enzyme_call.run_pass_pipeline(fns, source, pass_pipeline)
    810 if print_mlir:
    811     if type(print_mlir) != type(True):

ValueError: Failed to parse module
```

Reply:
```mlir
Module:
module {
  func.func @main(%arg0: tensor<4x1x28x28xf32>, %arg1: tensor<6x1x5x5xf32>, %arg2: tensor<6xf32>, %arg3: tensor<16x6x5x5xf32>, %arg4: tensor<16xf32>, %arg5: tensor<256x128xf32>, %arg6: tensor<128xf32>, %arg7: tensor<128x84xf32>, %arg8: tensor<84xf32>, %arg9: tensor<84x10xf32>, %arg10: tensor<10xf32>) -> tensor<4x10xf32> {
    %cst = stablehlo.constant dense<0.000000e+00> : tensor<84x4xf32>
    %cst_0 = stablehlo.constant dense<0.000000e+00> : tensor<128x4xf32>
    %cst_1 = stablehlo.constant dense<0.000000e+00> : tensor<8x8x16x4xf32>
    %cst_2 = stablehlo.constant dense<0.000000e+00> : tensor<24x24x6x4xf32>
    %cst_3 = stablehlo.constant dense<0xFF800000> : tensor<f32>
    %0 = stablehlo.transpose %arg0, dims = [3, 2, 1, 0] : (tensor<4x1x28x28xf32>) -> tensor<28x28x1x4xf32>
    %1 = stablehlo.transpose %arg1, dims = [3, 2, 1, 0] : (tensor<6x1x5x5xf32>) -> tensor<5x5x1x6xf32>
    %2 = stablehlo.transpose %arg3, dims = [3, 2, 1, 0] : (tensor<16x6x5x5xf32>) -> tensor<5x5x6x16xf32>
    %3 = stablehlo.reverse %1, dims = [0, 1] : tensor<5x5x1x6xf32>
    %4 = stablehlo.convolution(%0, %3) dim_numbers = [0, 1, f, b]x[0, 1, i, o]->[0, 1, f, b], window = {stride = [1, 1], pad = [[0, 0], [0, 0]], rhs_dilate = [1, 1]} {batch_group_count = 1 : i64, feature_group_count = 1 : i64} : (tensor<28x28x1x4xf32>, tensor<5x5x1x6xf32>) -> tensor<24x24x6x4xf32>
    %5 = stablehlo.broadcast_in_dim %arg2, dims = [2] : (tensor<6xf32>) -> tensor<24x24x6x4xf32>
    %6 = stablehlo.add %4, %5 : tensor<24x24x6x4xf32>
    %7 = stablehlo.compare  LT, %6, %cst_2 : (tensor<24x24x6x4xf32>, tensor<24x24x6x4xf32>) -> tensor<24x24x6x4xi1>
    %8 = stablehlo.select %7, %cst_2, %6 : tensor<24x24x6x4xi1>, tensor<24x24x6x4xf32>
    %9 = "stablehlo.reduce_window"(%8, %cst_3) <{padding = dense<0> : tensor<4x2xi64>, window_dilations = array<i64: 1, 1, 1, 1>, window_dimensions = array<i64: 2, 2, 1, 1>, window_strides = array<i64: 2, 2, 1, 1>}> ({
    ^bb0(%arg11: tensor<f32>, %arg12: tensor<f32>):
      %33 = stablehlo.maximum %arg11, %arg12 : tensor<f32>
      stablehlo.return %33 : tensor<f32>
    }) : (tensor<24x24x6x4xf32>, tensor<f32>) -> tensor<12x12x6x4xf32>
    %10 = stablehlo.reverse %2, dims = [0, 1] : tensor<5x5x6x16xf32>
    %11 = stablehlo.convolution(%9, %10) dim_numbers = [0, 1, f, b]x[0, 1, i, o]->[0, 1, f, b], window = {stride = [1, 1], pad = [[0, 0], [0, 0]], rhs_dilate = [1, 1]} {batch_group_count = 1 : i64, feature_group_count = 1 : i64} : (tensor<12x12x6x4xf32>, tensor<5x5x6x16xf32>) -> tensor<8x8x16x4xf32>
    %12 = stablehlo.broadcast_in_dim %arg4, dims = [2] : (tensor<16xf32>) -> tensor<8x8x16x4xf32>
    %13 = stablehlo.add %11, %12 : tensor<8x8x16x4xf32>
    %14 = stablehlo.compare  LT, %13, %cst_1 : (tensor<8x8x16x4xf32>, tensor<8x8x16x4xf32>) -> tensor<8x8x16x4xi1>
    %15 = stablehlo.select %14, %cst_1, %13 : tensor<8x8x16x4xi1>, tensor<8x8x16x4xf32>
    %16 = "stablehlo.reduce_window"(%15, %cst_3) <{padding = dense<0> : tensor<4x2xi64>, window_dilations = array<i64: 1, 1, 1, 1>, window_dimensions = array<i64: 2, 2, 1, 1>, window_strides = array<i64: 2, 2, 1, 1>}> ({
    ^bb0(%arg11: tensor<f32>, %arg12: tensor<f32>):
      %33 = stablehlo.maximum %arg11, %arg12 : tensor<f32>
      stablehlo.return %33 : tensor<f32>
    }) : (tensor<8x8x16x4xf32>, tensor<f32>) -> tensor<4x4x16x4xf32>
    %17 = stablehlo.transpose %16, dims = [3, 2, 1, 0] : (tensor<4x4x16x4xf32>) -> tensor<4x16x4x4xf32>
    %18 = stablehlo.reshape %17 : (tensor<4x16x4x4xf32>) -> tensor<4x256xf32>
    %19 = stablehlo.dot_general %arg5, %18, contracting_dims = [0] x [1], precision = [DEFAULT, DEFAULT] : (tensor<256x128xf32>, tensor<4x256xf32>) -> tensor<128x4xf32>
    %20 = stablehlo.broadcast_in_dim %arg6, dims = [0] : (tensor<128xf32>) -> tensor<128x4xf32>
    %21 = stablehlo.add %19, %20 : tensor<128x4xf32>
    %22 = stablehlo.compare  LT, %21, %cst_0 : (tensor<128x4xf32>, tensor<128x4xf32>) -> tensor<128x4xi1>
    %23 = stablehlo.select %22, %cst_0, %21 : tensor<128x4xi1>, tensor<128x4xf32>
    %24 = stablehlo.dot_general %arg7, %23, contracting_dims = [0] x [0], precision = [DEFAULT, DEFAULT] : (tensor<128x84xf32>, tensor<128x4xf32>) -> tensor<84x4xf32>
    %25 = stablehlo.broadcast_in_dim %arg8, dims = [0] : (tensor<84xf32>) -> tensor<84x4xf32>
    %26 = stablehlo.add %24, %25 : tensor<84x4xf32>
    %27 = stablehlo.compare  LT, %26, %cst : (tensor<84x4xf32>, tensor<84x4xf32>) -> tensor<84x4xi1>
    %28 = stablehlo.select %27, %cst, %26 : tensor<84x4xi1>, tensor<84x4xf32>
    %29 = stablehlo.dot_general %arg9, %28, contracting_dims = [0] x [0], precision = [DEFAULT, DEFAULT] : (tensor<84x10xf32>, tensor<84x4xf32>) -> tensor<10x4xf32>
    %30 = stablehlo.broadcast_in_dim %arg10, dims = [0] : (tensor<10xf32>) -> tensor<10x4xf32>
    %31 = stablehlo.add %29, %30 : tensor<10x4xf32>
    %32 = stablehlo.transpose %31, dims = [1, 0] : (tensor<10x4xf32>) -> tensor<4x10xf32>
    return %32 : tensor<4x10xf32>
  }
}
```

Reply:
Yeah that Module: at the top is not MLIR and should be be removed.

it should start with the
module {

Reply:
Relatedly why does reactant print that extra module name

Reply:
let me try fixing it on Reactant end